lib/event_store_client/connection/url_parser.rb

```diff
@@ -0,0 +1,145 @@
+# frozen_string_literal: true
+
+module EventStoreClient
+  module Connection
+    class UrlParser
```

Comment:
Hm... whats stops you to use https://ruby-doc.org/stdlib-2.7.0/libdoc/uri/rdoc/URI.html I have a feeling it can helps you here.